zkstack_cli/crates/zkstack/src/commands/chain/gateway/gateway_common.rs

```diff
@@ -0,0 +1,570 @@
+}
+
+impl MigrationDirection {
+    pub(crate) fn expected_notificaation(self) -> GatewayMigrationNotification {
+        match self {
+            Self::FromGateway => GatewayMigrationNotification::FromGateway,
+            Self::ToGateway => GatewayMigrationNotification::ToGateway,
+        }
+    }
+}
+
+/// Each migration to or from ZK gateway has multiple states it can be in.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub enum GatewayMigrationProgressState {
+    /// The state that represents that the migration has not yet started.
+    NotStarted,
+    /// The chain admin has sent the notification
+    NotificationSent,
+    /// The server has received the notification, but it is not yet ready for the migration.
+    NotificationReceived,
+    /// The server has received the notification and has no pending transactions
+    ServerReady,
+    /// The server is ready and the migration has started, but the server has not started sending transactions
+    /// to the new settlement layer yet.
+    AwaitingFinalization,
+    /// (Only for migrations from Gateway). The migration has been finalized on L1, but the user needs to execute it.
+    PendingManualFinalization,
+    /// The migration has finished.
+    Finished,
+}
+
+const MAX_SEARCHING_MIGRATION_TXS_INTERVAL: chrono::Duration = chrono::Duration::days(5);
+
+pub(crate) async fn get_migration_transaction(
+    sl_rpc_url: &str,
+    bridgehub_address: Address,
+    l2_chain_id: u64,
+) -> anyhow::Result<Option<H256>> {
+    let provider = get_ethers_provider(sl_rpc_url)?;
+    let sl_chain_id = provider.get_chainid().await?;
+
+    logger::info(format!(
+        "Searching for the migration transaction on SL {:#?}...",
+        sl_chain_id
+    ));
+
+    // Get the latest block so we know how far we can go
+    let mut search_upper_bound = provider
+        .get_block_number()
+        .await
+        .expect("Failed to fetch latest block")
+        .as_u64();
+
+    let bridgehub_contract = bridgehub_contract();
+
+    let max_interval_to_search = Utc::now() - MAX_SEARCHING_MIGRATION_TXS_INTERVAL;
+    let latest_event_log = loop {
+        let lower_bound = search_upper_bound.saturating_sub(DEFAULT_EVENTS_BLOCK_RANGE);
+
+        logger::info(format!(
+            "Checking block range: {}..={}",
+            lower_bound, search_upper_bound
+        ));
+
+        let filter = Filter::new()
+            .address(bridgehub_address)
+            .topic0(
+                bridgehub_contract
+                    .event("MigrationStarted")
+                    .unwrap()
+                    .signature(),
+            )
+            .from_block(lower_bound)
+            .topic1(u256_to_h256(U256::from(l2_chain_id)))
+            .to_block(search_upper_bound);
+
+        let result_logs = provider.get_logs(&filter).await?;
+        if !result_logs.is_empty() {
+            break result_logs.last().cloned();
+        }
+
+        if lower_bound == 0 {
+            break None;
+        }
```

Comment:
It's still zero let's go back not so far from

Reply:
But there is a limit on time of 5 days, I think having limit on time is better than the one based on blocks